multimap/src/main/java/org/infinispan/multimap/impl/HashMapBucket.java

```diff
@@ -116,22 +119,23 @@ public boolean containsKey(K key) {
    /**
     * We do not use the original replace method here. Our implementation allows to create and delete entries.
     */
-   public boolean replace(K key, V expected, V replacement) {
+   public HashMapBucket<K, V> replace(K key, V expected, V replacement) {
+      Map<MultimapObjectWrapper<K>, V> copied = new HashMap<>(values);
```

Comment:
Same again it would be better to iterate over the original contents and copy in and skip the one that matches for the replace.

Reply:
We may want to re-evaluate all the places we are making this copy to see if it can be optimized.

Reply:
I've revisited the other buckets to check and updated the "remove" methods to avoid the full copy. However, I've kept the methods that add new entries to the bucket. I guess we can't escape the copy in these cases.